Fix #1090 Consolidate/Simplfy ServerOpts

hkeebler opened the pull request:
Uses property overrides instead of command line options and includes GC safemode option.

This includes:
- properties for <servertype>.bind.address added to Property.java with defaults
- property for GC safemode
- deprecation of safemode command line option
- removal of --address command line option (deprecating would have made for some ugly code - could be revisited if desired)
- addition of a GC safemode IT test (could not find one already out there) .   Could be improved - copied from GarbargeCollectorIT.  Trying to find a way to override a property (not from the command line) was difficult.

This did not address merging the host and port properties together or discerning bind addresses from advertised addresses.  I'm not sure what the difference really is.  Monitor was the only server that used an "adveristedHost" which was the bind address.
Note that the override properties processing as of now does not take into account boolean type properties with no value assigned.  Maybe a new ticket as well as documentation updates.

hkeebler commented on core/src/main/java/org/apache/accumulo/core/conf/Property.java:
```diff
@@ -515,6 +519,8 @@
   GC_CYCLE_DELAY("gc.cycle.delay", "5m", PropertyType.TIMEDURATION,
       "Time between garbage collection cycles. In each cycle, old RFiles or write-ahead logs "
           + "no longer in use are removed from the filesystem."),
+  GC_HOST("gc.bind.address", "0.0.0.0", PropertyType.HOSTLIST,
```
Used PropertyType.HOSTLIST because it already did some validation even through these would only be a single entry.

ctubbsii replied:
Probably okay for a first pass, but probably better to create a new type. These need to specifically be IP addresses (IPv4 or IPv6) and *not* hostnames. They would also be a single entry (and might permit a port to be specified... possibly).

hkeebler replied:
The HOSTLIST validation includes checking for port which is why I thought to go ahead and use it but did not realize that ip addresses were required.  A new type would be a good idea.

hkeebler commented on server/base/src/main/java/org/apache/accumulo/server/AbstractServer.java:
```diff
@@ -59,6 +56,7 @@ protected AbstractServer(String appName, ServerOpts opts, String[] args) {
    * Run this server in a main thread
    */
   public void runServer() throws Exception {
+    TraceUtil.enableServerTraces(getHostname(), applicationName, context.getConfiguration());
```
This was moved from the constructor due to the hostname not being resolved at time abstract constructor was called.

ctubbsii replied:
Probably fine... as long as the remaining body of the constructor doesn't try to trace anything.

Also, instead of moving this, we could try to make the setting of the hostname more standardized by moving it into the AbstractServer constructor itself.

hkeebler replied:
Yes.  It was very clean before when the address was more generic but using properties made it more difficult to manage in the absract server without having a long switch statement for each server type.  The property could use the "appname" which is already passed in as a prefix (I believe the naming convention is this already).   That actually may work..hmmm.

ctubbsii replied:
The appname may work... but that might also be a user-configurable parameter, intended for metrics and/or logging configuration... so it might not consistently represent the configuration parameter name. A switch statement or another required method on the abstract server base class to provide the config prefix based on the server type might be needed.

hkeebler commented on server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java:
```diff
@@ -20,6 +20,7 @@
+@Deprecated
 public class ServerOpts extends ConfigOpts {
 
   @Parameter(names = {"-a", "--address"}, description = "address to bind to")
```
This could be deleted.  It is not being used unless making this a deprecated option is desired.

ctubbsii replied:
This could be replaced with a simple check of the args list for any `-a` or `--address` strings, print an error message with the new replacement option. Could also permit this, with a warning, and use its value to override whatever the configuration parsing results in... but that would add some complexity that might not be worth it (as you've already alluded to in your PR comment body).

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/cli/ConfigOpts.java:
```diff
@@ -45,6 +47,14 @@ public synchronized String getPropertiesPath() {
+  public static class DeprecatedOption implements IParameterValidator {
+    @Override
+    public void validate(String option, String value) throws ParameterException {
+      System.out.println(
```
Should this use `System.err` instead? Or better yet, perhaps `log.warn`, so the behavior is user-configurable within the user's logging settings?

I'd hate for the printing of this to interfere with any automation/scripting a user might be doing, or anything else that might be capturing STDOUT.

hkeebler replied:
Understand but note that the GC does the same although I guess if using safemode then stdout is expected.  There could be future deprecated options that do not involve any STDOUT.   The --help will print the deprecation info to the console so if the option is actually used the warning could just be logged, log.warning.  I"m good with that.
```
 if (safemode) {
        System.out.println("SAFEMODE: There are " + confirmedDeletes.size()
            + " data file candidates marked for deletion.%n"
            + "          Examine the log files to identify them.%n");
```

ctubbsii replied:
Since the help info already has the deprecation info, logging a warning when used here is fine.

If I remember correctly, I did argue against using STDOUT at the time it was added to the gc safemode, but it was make clear to me that safemode is typically run in a manual way, not scripted with a user's typical launch scripts, so STDOUT has specific utility for the user running it in safemode (though it could also be STDERR in that case, too). In general, we should use less `System.{out,err,exit,<probably others>}`, but there's always going to be exceptions to best practices.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/conf/Property.java:
```diff
@@ -230,6 +230,8 @@
   // properties that are specific to master server behavior
   MASTER_PREFIX("master.", null, PropertyType.PREFIX,
       "Properties in this category affect the behavior of the master server"),
+  MASTER_HOST("master.bind.address", "0.0.0.0", PropertyType.HOSTLIST,
```
Regarding this naming convention, should these be `*.bind.address` or `*.address.bind`? What do you think? `master.bind.address` reads more naturally, but `master.address.bind` supports logically grouping address-related configs in a configuration hierarchy, which can be better for parsing, and more easily incorporate future changes.

An example of a future change which could benefit from the `*.address.bind` form could include work that address an explicit public advertisement address, in the case of the bind address not being publicly reachable (in the case of more complicated networking setups, such as those on some cloud services' infrastructure):

Example:
```ini
; to add a publicly reachable service address (the one recorded in ZooKeeper for clients trying
; to reach this server), in addition to the bind address we would need a public address:
master.bind.address = 172.16.0.1
master.public.address = 203.0.113.1

; alternative naming convention
master.address.bind = 0.0.0.0
master.address.public = 203.0.113.1 # or accumulo-master.example.com b/c this can be a hostname

; another possible future change could include ports to replace `*.port.client`,
; so we reduce the number of config options, and put address information all in one place
master.address.bind = 0.0.0.0:9999
master.address.public = accumulo-master.example.com:50999
```

I'm leaning towards the `<service>.address.<addressType>` naming convention, rather than the `<service>.<addressType>.address` convention that you currently have here, but am open to discussion... because naming is hard. What do you think?

hkeebler replied:
Naming is always hard.  Your suggestion to make it *.address.bind works for me.  I'm surprised there were no comments on the Property name MASTER_HOST - I like short if I can get away with it but maybe it should be MASTER_ADDRESS or even MASTER_BIND_ADDRESS

ctubbsii replied:
I'm not concerned about the enum name at all, since that's not user-facing. It's internal only.

At one point I thought about making the property names derived from the enum names (basically, make lowercase, and replace underscores with dots, all in code). I did this at one point for some enums in our MapReduce code. But, it would require us to get pretty strict about enum naming in Property.java, and at this point, I don't think it's worth it.

Long story short: the enum can be whatever. :smiley_cat:

ctubbsii commented on server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java:
```diff
@@ -108,11 +108,11 @@
-    boolean verbose = false;
-    @Parameter(names = {"-s", "--safemode"}, description = "safe mode will not delete files")
+  static class GCOpts extends ConfigOpts {
+    @Parameter(names = {"-s", "--safemode"}, validateWith = DeprecatedOption.class,
```
Will this cause this to show up in any automated help documentation? It might be easier to leave this in the ServerOpts file for now, with your new warning.

hkeebler replied:
Yes it is printed with a warning saying that it is deprecated:
 ```
@Parameter(names = {"-s", "--safemode"}, validateWith = DeprecatedOption.class,
        description = "This option is DEPRECATED and replaced with property gc.safemode=true. "
            + "Safe mode will list files but not delete them")
```
This was  never in the ServerOpts - only specific to GC.

ctubbsii commented on server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java:
```diff
@@ -620,7 +620,7 @@ private HostAndPort startStatsService() {
     }
     int[] port = getConfiguration().getPort(Property.GC_PORT);
-    HostAndPort[] addresses = TServerUtils.getHostAndPorts(this.opts.getAddress(), port);
+    HostAndPort[] addresses = TServerUtils.getHostAndPorts(getHostname(), port);
```
I'm not sure this is right. There is a difference between the address used for the bind address, and the hostname and port that was previously returned from `getHostname()` that was stored in ZooKeeper for the service address.

hkeebler replied:
The AbstractServer set hostname=opts.getAddress() and provided a method getHostname() used by the extending server to get their host address.  Note that opts.getAddress() defaults to 0.0.0.0 when no address was provided on the command line.

ctubbsii replied:
Okay, so I was confused. I had thought `getHostname()` returned the name lookup from the IP address, but apparently that is done in `TServerUtils` much, much later. So, you're right... this is effectively the same... and there's no reason to use the opts here when `super.getHostname()` is sufficient. The method is, perhaps, badly named and misleading, but that can be addressed when the `--address` stuff is addressed, in future.

hkeebler replied:
I thought about changing it to getAddress() but as you said ...later.

ctubbsii commented on server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java:
```diff
@@ -671,6 +671,11 @@ static boolean isDir(String delete) {
   }
 
+  @Override
+  public String getHostname() {
```
I believe this is supposed to return the service address stored in ZooKeeper, not the bind address. It might make sense to move this back to the abstract super class.

hkeebler replied:
See previous comment but if the hostname assignment is moved back to the Abstract class then what is going on may be more clear.

hkeebler commented:
Added the hostname assignment back into the AbstractServer.  I don't love that the appName is a string assigned and hardcoded in each server class but it works for now.  It probably would be good to create an ApplicatioName enum for these (another ticket).
```java
 protected AbstractServer(String appName, ConfigOpts opts, String[] args) {
    this.log = LoggerFactory.getLogger(getClass().getName());
    this.applicationName = appName;
    opts.parseArgs(appName, args);
    var siteConfig = opts.getSiteConfiguration();
    this.hostname = siteConfig.get(appName+ Property.SERVER_ADDRESS_BIND_SUFFIX); <====
    context = new ServerContext(siteConfig);
    SecurityUtil.serverLogin(siteConfig);
    log.info("Version " + Constants.VERSION);
    log.info("Instance " + context.getInstanceID());
    ServerUtil.init(context, appName);
    this.metricsSystem = Metrics.initSystem(getClass().getSimpleName());
    TraceUtil.enableServerTraces(hostname, appName, context.getConfiguration());
    if (context.getSaslParams() != null) {
      // Server-side "client" check to make sure we're logged in as a user we expect to be
      context.enforceKerberosLogin();
    }
  }
```

Also added a deprecation property override so it may make it easier to deprecate the --address option .  Still reviewing.

```
 /**
   * Performs special processing of deprecated options.
   * Could be cleaned up periodically as deprecations are deleted
   * @param option
   * @param value
   */
  private static void processDeprecation(String option, String value){
    if ("-s".equals(option)){
      deprecated_overrides.add(Property.GC_SAFEMODE.getKey()+"=true");
    }
  }
```

ctubbsii commented:
> Also, if possible could you provide a use case for starting a server from the command line and specifying a different address?

The most common use case is to specify `0.0.0.0` as the bind address, but in ZooKeeper, it would still select whatever the `InetAddress.getHostName()` was for storing in ZooKeeper (because `0.0.0.0` can't be used... it's not reachable).

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/cli/ConfigOpts.java:
```diff
@@ -104,4 +121,14 @@ public void parseArgs(String programName, String[] args, Object... others) {
+   * deprecations are deleted
+   */
+  private static void processDeprecation(String option, String value) {
+    if ("-s".equals(option)) {
```
`--address` is definitely going to be its own beast. I wouldn't worry about it here.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/cli/ConfigOpts.java:
```diff
@@ -57,6 +69,9 @@ public synchronized String getPropertiesPath() {
   private List<String> overrides = new ArrayList<>();
 
+  // holds property overrides that may be generated from deprecation:
+  private static List<String> deprecated_overrides = new ArrayList<>();
```
Those legacy options are more for client-specific config. But, throwing the exception might be a good model to follow instead of trying to translate.

hkeebler replied:
So...I believe we are saying that the best path forward at this time is to do something similar to how ClientOpts is handling legacy (or dropped) options.  No deprecation and auto setting properties.  This definitely simplifies things and keeps the code tidier.   Since these are command line options, dropping them instead of "deprecating" them is not so bad since they are easily updated by the customer vs. "deprecated" java code.  The worst case being that customer generated scripts may need to be updated.

ctubbsii replied:
> This definitely simplifies things and keeps the code tidier. Since these are command line options, dropping them instead of "deprecating" them is not so bad since they are easily updated by the customer vs. "deprecated" java code.

I think you just convinced me that this is the right course of action.

hkeebler replied:
farout! ;)

hkeebler commented:
Latest commit addresses the following suggestions:
- dropping option over deprecation
- using ClientOpts legacyOpts model (overkill for just the safemode option but sets it up for future changes like the --address )

>  Perhaps if we kept a static map of known replacements, we could provide a more specific warning message here.
` value = Property.isValidBooleanPropertyKey(key) ? "true" : "";`

ctubbsii commented on server/gc/src/test/java/org/apache/accumulo/gc/SimpleGarbageCollectorTest.java:
```diff
@@ -94,6 +94,7 @@ public void testInit() {
     assertTrue(gc.isUsingTrash());
     assertEquals(1000L, gc.getStartDelay());
     assertEquals(2, gc.getNumDeleteThreads());
+    assertFalse(gc.inSafeMode()); //false by default
```
When building the PR, the formatter added a space between `//` and `false by default`. Probably good to run the formatter before merging to avoid a second formatting-only commit to fix.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/cli/ConfigOpts.java:
```diff
@@ -96,6 +117,19 @@ public synchronized SiteConfiguration getSiteConfiguration() {
+      // prints error to console if ran from the command line otherwise there is no way to know that
+      // an error occurred
+      System.err.println(errmsg.append("See '-o' property override option").toString());
+      throw new IllegalArgumentException(errmsg.toString());
```
Forget this comment... you probably do still need it right now... but only because of the problem I saw in #1349 .

ctubbsii commented on server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java:
```diff
@@ -258,12 +250,10 @@ public boolean getCandidates(String continuePoint, List<String> result)
-        }
+      if (inSafeMode()) {
+        System.out.println("SAFEMODE: There are " + confirmedDeletes.size()
+            + " data file candidates marked for deletion.%n"
```
That output looks great. I wasn't thinking about providing the ZooKeeper node, just literally the word "ZooKeeper", but this is fine.
